File: Fourth term/Fibers/Java/Process.java
package com.lab;

import java.util.ArrayList;
import java.util.Random;

public class Process {

    private static final int longPauseBoundary = 10000;
    private static final int shortPauseBoundary = 100;
    private static final int workBoundary = 1000;
    private static final int intervalsAmountBoundary = 10;
    private static final int priorityLevelsNumber = 10;

    private final ArrayList<Integer> workIntervals = new ArrayList<>();
    private final ArrayList<Integer> pauseIntervals = new ArrayList<>();

    private int priority;
    private int totalDuration;
    private int activeDuration;

    public ArrayList<Integer> getWorkIntervals() {
        return workIntervals;
    }

    public int getPriority() {
        return priority;
    }

    public void setPriority(int priority) {
        this.priority = priority;
    }

    public int getTotalDuration() {
        return activeDuration + sumOfElements(pauseIntervals);
    }

    public int getActiveDuration() {
        return sumOfElements(workIntervals);
    }

    public Process() {
        Random random = new Random();
        int amount = random.nextInt(intervalsAmountBoundary);

        for (int i = 0; i < amount; i++) {
            workIntervals.add(random.nextInt(workBoundary));
            pauseIntervals.add(random.nextInt(
                    random.nextDouble() > 0.9
                            ? longPauseBoundary
                            : shortPauseBoundary));
        }
        priority = random.nextInt(priorityLevelsNumber);
    }

    public void run() throws InterruptedException {
        synchronized (this) {
            for (int i = 0; i < workIntervals.size(); i++) {
                Thread.sleep(workIntervals.get(i)); // work emulation

                long pauseBeginTime = System.currentTimeMillis();
                do {
                    ProcessManager.processManagerSwitch(false);
                }
                while ((System.currentTimeMillis() - pauseBeginTime) < pauseIntervals.get(i)); // I/O emulation*/
            }

            ProcessManager.processManagerSwitch(true);
        }
    }

    private int sumOfElements(ArrayList<Integer> elements) {
        int sum = 0;
        for (int element : elements) {
            sum += element;
        }
        return sum;
    }
}

File: IgorKorobitsyn/Block1Task1/src/main/kotlin/Process.kt
package fibers

import kotlin.random.Random
import kotlin.random.nextUInt

class Process(private val pm: ProcessManager)  {


    private val LONG_PAUSE_BOUNDARY = 10000
    private val SHORT_PAUSE_BOUNDARY = 100
    private val WORK_BOUNDARY = 1000
    private val INTERVALS_AMOUNT_BOUNDARY = 10
    private val PRIORITY_LEVELS_NUMBER = 10

    private var workIntervals: ArrayList<Int> = arrayListOf()
    private var pauseIntervals: ArrayList<Int> = arrayListOf()

    private var priority = 0
    private var totalDuration = 0
    private var activeDuration = 0

    fun getWorkIntervals(): List<Int> {
        return workIntervals
    }

    fun getPriority(): Int {
        return priority
    }

    fun setPriority(newPriority: Int) {
        priority = newPriority
    }

    fun getTotalDuration(): Int {
        totalDuration = activeDuration + pauseIntervals.sum()
        return totalDuration
    }

    fun getActiveDuration(): Int {
        activeDuration = workIntervals.sum()
        return activeDuration
    }

    init {
        var random: Random = Random
        var amount: Int = random.nextInt(INTERVALS_AMOUNT_BOUNDARY)

        for (i in 1..amount){
            workIntervals.add(random.nextInt(WORK_BOUNDARY))
            pauseIntervals.add(random.nextInt(
                if (random.nextDouble() > 0.9) {LONG_PAUSE_BOUNDARY}
                else {SHORT_PAUSE_BOUNDARY})
            )
        }

        priority = random.nextInt(PRIORITY_LEVELS_NUMBER)

        println("Process of priority ${priority} was created")
    }

    fun run(){
        for (i in 1..workIntervals.size) {
            Thread.sleep(workIntervals[i].toLong())

            var pauseBeginTime = System.currentTimeMillis()

            do {
                pm.switch(false)
            }while ((System.currentTimeMillis() - pauseBeginTime) < pauseIntervals[i])
        }

        pm.switch(true)
    }
}